egui/src/widgets/plotting/items.rs

```diff
@@ -0,0 +1,214 @@
+
+/// A horizontal line in a plot, filling the full width
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct HLine {
+    pub(crate) y: f64,
+    pub(crate) stroke: Stroke,
+}
+
+impl HLine {
+    pub fn new(y: impl Into<f64>, stroke: impl Into<Stroke>) -> Self {
+        Self {
+            y: y.into(),
+            stroke: stroke.into(),
+        }
+    }
+}
+
+/// A vertical line in a plot, filling the full width
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct VLine {
+    pub(crate) x: f64,
+    pub(crate) stroke: Stroke,
+}
+
+impl VLine {
+    pub fn new(x: impl Into<f64>, stroke: impl Into<Stroke>) -> Self {
+        Self {
+            x: x.into(),
+            stroke: stroke.into(),
+        }
+    }
+}
+
+// ----------------------------------------------------------------------------
+
+enum Generator {
+    /// Describes a function y = f(x) with an optional range for x and a number of points.
+    Explicit(Box<dyn Fn(f64) -> f64>, Option<RangeInclusive<f64>>, usize),
+    /// Describes a function (x,y) = f(t) with a range for t and a number of points.
+    Parametric(Box<dyn Fn(f64) -> (f64, f64)>, RangeInclusive<f64>, usize),
+}
+
+// ----------------------------------------------------------------------------
+
+/// A series of values forming a path.
+pub struct Curve {
+    pub(crate) values: Vec<Value>,
+    generator: Option<Generator>,
+    pub(crate) bounds: Bounds,
+    pub(crate) stroke: Stroke,
+    pub(crate) name: String,
+}
+
+impl Curve {
+    fn empty() -> Self {
+        Self {
+            values: Vec::new(),
+            generator: None,
+            bounds: Bounds::NOTHING,
+            stroke: Stroke::new(2.0, Color32::TRANSPARENT),
+            name: Default::default(),
+        }
+    }
+
+    pub fn from_values(values: Vec<Value>) -> Self {
+        let mut bounds = Bounds::NOTHING;
+        for value in &values {
+            bounds.extend_with(value);
+        }
+        Self {
+            values,
+            bounds,
+            ..Self::empty()
+        }
+    }
+
+    pub fn from_values_iter(iter: impl Iterator<Item = Value>) -> Self {
+        Self::from_values(iter.collect())
+    }
+
+    /// Draw a curve based on a function `y=f(x)`, an optional range for x and the number of points.
+    pub fn from_explicit_callback(
+        function: impl Fn(f64) -> f64 + 'static,
+        range: Option<RangeInclusive<f64>>,
+        points: usize,
+    ) -> Self {
+        let generator = Generator::Explicit(Box::new(function), range, points);
+        Self {
+            generator: Some(generator),
+            ..Self::empty()
+        }
+    }
+
+    /// Draw a curve based on a function `(x,y)=f(t)`, a range for t and the number of points.
+    pub fn from_parametric_callback(
+        function: impl Fn(f64) -> (f64, f64) + 'static,
+        t_range: RangeInclusive<f64>,
+        points: usize,
+    ) -> Self {
+        let generator = Generator::Parametric(Box::new(function), t_range, points);
```

Comment:
The parametric function always computes all the points, so there is not reason to defer that computation to later as far as I can tell.

If you compute values here instead and pass to `Self::from_values` we get default bounds and also a speedup, because `function` doesn't need to be boxed (and doesn't need to be `'static` either)